Fix https://github.com/NuGet/Home/issues/1816

jainaashish opened the pull request:
Removing versions which don't satisfy allowedVersions before picking the highest version

Fixes: https://github.com/NuGet/Home/issues/1816

dnfclas commented:
Hi __@jainaashish__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

jainaashish commented:
@emgarten @alpaix

alpaix commented:
@jainaashish Can you add test(s) to illustrate the issue and the fix?

emgarten commented:
This seems correct. I would double check that no tests are failing, and verify that if there are no packages in the allowed range that it gives a good error message for that instead of saying that it cannot find the package at all. 

In the past I've found that small changes to order here can change the many errors displayed.

jainaashish commented:
So i've updated the PM UI to consider allowedVersion range defined in package.config to show updates available. Also added/updated tests for the issue fixed.

new UX will look like this:
![image](https://cloud.githubusercontent.com/assets/18219758/15522097/d56c78b8-21c4-11e6-9d4a-cd27f758fb24.png)

jainaashish commented:
@emgarten @alpaix @rrelyea @yishaigalatzer @harikm86

harikmenon commented:
We should not allow update of a package to a version that is outside of the bounds of the version constraints. I can see value in selecting a version that is below the version constraints not satisfied header so the user can check out release notes or whatever, but we shouldn't allow the update. We should disable the button and possibly have a message explaining why that is the case.

yishaigalatzer commented:
There are cases wants to override the constraints, so we need to consider that blocking updates will cause undesired side effects (which is why we already have the ignore dependencies check box).

So please take time and design the experience end to end. Justin can help with advice

jainaashish commented:
As discussed with Hari and Justin, we finalized that we'll disable blocked versions to update or install until user changes dependency behavior. If dependency behavior is changed to ignore, then we'll enable these blocked versions to update or install.

emgarten commented on src/NuGet.Clients/PackageManagement.PowerShellCmdlets/NuGetPowerShellBaseCommand.cs:
```diff
@@ -569,13 +569,13 @@ protected IEnumerable<IPackageSearchMetadata> GetPackagesFromRemoteSource(string
         protected async Task<IEnumerable<IPackageSearchMetadata>> GetPackagesFromRemoteSourceAsync(string packageId, bool includePrerelease)
         {
-            var metadataProvider = new MultiSourcePackageMetadataProvider(PrimarySourceRepositories, optionalLocalRepository: null, optionalGlobalLocalRepository: null, logger: Common.NullLogger.Instance);
+            var metadataProvider = new MultiSourcePackageMetadataProvider(PrimarySourceRepositories, optionalLocalRepository: null, optionalGlobalLocalRepository: null, projects: new NuGetProject[] { Project }, isSolution:false, logger: Common.NullLogger.Instance);
```
nit: format the document, the spacing looks off here

emgarten commented on src/NuGet.Clients/PackageManagement.UI/Feeds/MultiSourcePackageMetadataProvider.cs:
```diff
@@ -75,8 +90,34 @@ internal class MultiSourcePackageMetadataProvider : IPackageMetadataProvider
+                {
+                    // get all package references for all the projects and cache locally
+                    var packageReferencesTasks = _projects.Select(project => project.GetInstalledPackagesAsync(cancellationToken));
+                    _packageReferences = (await Task.WhenAll(packageReferencesTasks)).SelectMany(p => p).Where(p => p != null);
```
Can these actually be null?

jainaashish replied:
yes, first time packageReferences will be null. And it will be initialized which will be cached and reused.

emgarten commented on src/NuGet.Clients/PackageManagement.UI/Feeds/MultiSourcePackageMetadataProvider.cs:
```diff
@@ -75,8 +90,34 @@ internal class MultiSourcePackageMetadataProvider : IPackageMetadataProvider
+                if (allowedVersionsRange.Any())
+                {
+                    // Find a common range which satisfies all the version ranges across projects
+                    allowedVersions = VersionRange.FindCommonRange(allowedVersionsRange);
```
This looks like a great way to handle multiple ranges, but this is inside the if (!_isSolution) block, so I don't see how they could ever contain more than 1 range?

jainaashish replied:
yeah initially i didn't had that solution check, but now it will only have single range. will update it.

emgarten commented on src/NuGet.Clients/PackageManagement.UI/Feeds/MultiSourcePackageMetadataProvider.cs:
```diff
@@ -75,8 +90,34 @@ internal class MultiSourcePackageMetadataProvider : IPackageMetadataProvider
         public async Task<IPackageSearchMetadata> GetLatestPackageMetadataAsync(PackageIdentity identity,
             bool includePrerelease, CancellationToken cancellationToken)
         {
+            VersionRange allowedVersions = null;
```
This should default to the _All_ range instead of null.

`VersionRange allowedVersions = VersionRange.All;`

emgarten commented on src/NuGet.Clients/PackageManagement.UI/Feeds/SourceRepositoryExtensions.cs:
```diff
@@ -136,7 +136,10 @@ public static Task<SearchResult<IPackageSearchMetadata>> SearchAsync(this Source
 
-            var highest = packages?
+            // filter packages based on allowed versions
+            var updatedPackages = allowedVersions != null ? packages.Where(p => allowedVersions.Satisfies(p.Identity.Version)) : packages;
```
This null check can be removed if VersionRange.All is used

emgarten commented on src/NuGet.Clients/PackageManagement.UI/Models/DetailControlModel.cs:
```diff
@@ -249,18 +262,8 @@ protected void SelectVersion()
-                SelectedVersion = versionToSelect;
-            }
+            // it should always select the top version from versions list to install or update
+            SelectedVersion = _versions[0];
```
Doesn't this need to check that there are versions? The check removed here was to fix major bug.

jainaashish replied:
It already checks for versions available. With disabled version, it just needs to further check for valid versions which i'm adding with new changes.

emgarten commented on src/NuGet.Clients/PackageManagement.UI/Models/PackageItemListViewModel.cs:
```diff
@@ -260,6 +262,9 @@ public void UpdatePackageStatus(IEnumerable<PackageIdentity> installedPackages)
                     var packageVersions = await GetVersionsAsync();
+
+                    // filter package versions based on allowed versions in pacakge.config
+                    packageVersions = AllowedVersions != null ? packageVersions.Where(v => AllowedVersions.Satisfies(v.Version)) : packageVersions;
```
Use VersionRange.All to ensure this is never null

emgarten commented on src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs:
```diff
@@ -376,6 +376,146 @@ public static VersionRange Combine(IEnumerable<VersionRange> ranges, IVersionCom
+        /// <summary>
+        /// Returns the greatest common range that satisfies all given ranges.
+        /// </summary>
+        public static VersionRange FindCommonRange(IEnumerable<VersionRange> ranges, IVersionComparer comparer)
```
How about CommonSubSet?

emgarten commented on test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeOperationTests.cs:
```diff
@@ -235,5 +235,78 @@ public void VersionRangeSet_CombineMultipleRanges()
+        [InlineData("(0.0.0, 0.0.0)", "[1.0.0, 3.0.0)", "[4.0.0, 5.0.0)")]
+        [InlineData("[1.5.0, 2.0.0]", "[1.0.0, 2.0.0]", "[1.5.0, 3.0.0]")]
+        [InlineData("[1.0.0, 1.5.0)", "[1.0.0, 1.5.0)", "[, ]")]
+        public void VersionRangeSet_FindCommonRange(string expected, string rangeA, string rangeB)
```
Add tests that include since version ranges such as [1.0.0] and (1.0.0)

emgarten commented:
AllowedVersions has never been a first class feature. This is a large change which still makes it only a partial feature. It still doesn't have a way to modify or add the ranges from the UI or even the console. I'm not sure that the increased UI complexity here will benefit the number advanced users manually editing their packages.config to add this attribute.

Looking at the original bug the user is using nuget.exe, there haven't been any user requests for this.

:watch:

jainaashish commented:
As discussed offline, we'll continue with the fix but will make sure we've a manual UI test added to test the feature. Besides I've already added test to cover original bug for pruning versions based on allowedVersions before choosing the highest available.

emgarten commented on src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs:
```diff
@@ -376,6 +376,146 @@ public static VersionRange Combine(IEnumerable<VersionRange> ranges, IVersionCom
+
+            // Remove zero width ranges. Ex: (1.0.0, 1.0.0)
+            // This includes VersionRange.None and any other ranges that satisfy zero versions
+            ranges = ranges.Where(range => HasValidRange(range));
```
Invalid ranges should cause this method to return None instead of being ignored. Add a test for this also

emgarten commented on src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs:
```diff
@@ -376,6 +376,146 @@ public static VersionRange Combine(IEnumerable<VersionRange> ranges, IVersionCom
+                }
+
+                // at last just check the final lowest n highest versions
+                if(lowest != null && highest != null)
```
nit: formatting

emgarten commented on src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs:
```diff
@@ -376,6 +376,146 @@ public static VersionRange Combine(IEnumerable<VersionRange> ranges, IVersionCom
+                            }
+                            // lowerCompare < 0 out of the current range, this is a no-op
+
+                            // compare new lowest with existing highest, if it's going out of tha range
```
_the_ range

emgarten commented on src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs:
```diff
@@ -376,6 +376,146 @@ public static VersionRange Combine(IEnumerable<VersionRange> ranges, IVersionCom
+                            // compare new lowest with existing highest, if it's going out of tha range
+                            if (highest != null)
+                            {
+                                lowerCompare = comparer.Compare(highest, lowest);
```
This is checked at the end, why is it checked here also?

emgarten commented on src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs:
```diff
@@ -376,6 +376,146 @@ public static VersionRange Combine(IEnumerable<VersionRange> ranges, IVersionCom
+        /// <summary>
+        /// Returns the greatest common range that satisfies all given ranges.
+        /// </summary>
+        public static VersionRange CommonSubSet(IEnumerable<VersionRange> ranges, IVersionComparer comparer)
```
I think this entire method could be simplified, finding the innermost set is easier than finding the outermost set (combine). In combine this pattern is used to verify there are no gaps.

You should able to: 
1. Find the highest lower bound  `ranges.Where(r => r.MinVersion != null).Max(r => r.MinVersion)`
2. Find the lowest higher bound
3. Set the bound to exclusive if any matches on that version have it set to exclusive
4. Check if the range is valid

Here if there are any gaps, for example  [1.0.0, 2.0.0]  [4.0.0, 5.0.0]  the result will be [4.0.0, 2.0.0] which is invalid.

jainaashish replied:
make sense, I'll simplify it.

emgarten replied:
Looks great :smile:

emgarten commented on test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeOperationTests.cs:
```diff
@@ -235,5 +235,80 @@ public void VersionRangeSet_CombineMultipleRanges()
+        [InlineData("[1.0.0, 1.5.0)", "[1.0.0, 1.5.0)", "[, ]")]
+        [InlineData("[1.0.0, 1.0.0]", "[1.0.0]", "(1.0.0)")]
+        [InlineData("(0.0.0, 0.0.0)", "[1.0.0]", "[2.0.0]")]
+        public void VersionRangeSet_CommonSubSet(string expected, string rangeA, string rangeB)
```
This could use more tests around inclusive/exclusive boundaries

emgarten commented on src/NuGet.Clients/PackageManagement.UI/Models/PackageDetailControlModel.cs:
```diff
@@ -78,8 +78,16 @@ protected override void CreateVersions()
-            var latestStableVersion = allVersions.FirstOrDefault(v => !v.IsPrerelease);
+
+            // null, if no version constraint defined in package.config
+            var allowedVersions = _projectVersionRangeDict.Select(kvp => kvp.Value).FirstOrDefault();
```
var allowedVersions = _projectVersionRangeDict.Select(kvp => kvp.Value).FirstOrDefault() ?? VersionRange.All

and remove the null check below to simplify this

emgarten commented:
:watch: simplify CommonSubSet and add more tests

Commit:
Fix https://github.com/NuGet/Home/issues/1816

Removing versions whish don't satisfy allowedVersions before picking the highest version

emgarten commented:
LGTM :rocket: :shipit:

jainaashish commented:
Merged 2cc8fc0b69c3189ccc6ed2234d6ac3dcf4945bb2
